drivers/adc/adc_smartbond_gpadc.c

```diff
@@ -243,20 +280,83 @@ static int adc_smartbond_read_async(const struct device *dev,
 	struct adc_smartbond_data *data = dev->data;
 	int error;
 
+	gpadc_smartbond_pm_policy_state_lock_get(data);
 	adc_context_lock(&data->ctx, true, async);
 	error = start_read(dev, sequence);
 	adc_context_release(&data->ctx, error);
+	gpadc_smartbond_pm_policy_state_lock_put(data);
```

Comment:
We should call `gpadc_smartbond_pm_policy_state_lock_put` when the reading has completed, shouldn't we?

Reply:
Requested ADC conversions should be done once `start_read` returns. Looking at the mentioned routine we block by calling `adc_context_wait_for_completion`. Unblocking takes place in `adc_smartbond_isr` by calling `adc_context_on_sampling_done`.  As such, PM policy constraints could be done before/after calling `start_read`.

Reply:
@ioannis-karachalios If that was the case, we wouldn't have a correct asynchronous API implementation.
`adc_context_wait_for_completion()` returns early without taking a semaphore when we make an asynchronous reading.
Same stands for `adc_context_release()` in case of asynchronous read. It returns early without releasing `ctx->lock`.

Reply:
You are right. The API naming is sort of misleading. As such, PM policy constraints should be released form within ISR when all requested channels are sampled, that should be right after `adc_context_on_sampling_done`.